### ISSUE TITLE
Clinica Run T1-Linear Fails Due to FileNotFoundError in N4BiasFieldCorrection Node

### ISSUE THREAD

**Issue by mtmmu88** [aramis-lab/clinica#998]
Dear Clinica Developers,

I've encountered an issue when running the t1-linear pipeline using Clinica on my Linux machine. Specifically, I get a FileNotFoundError related to the 'bias_image' in the N4BiasFieldCorrection node. Below are the details:

Command Executed:
clinica run t1-linear /home/zzc/Output /home/zzc/predict_2
Error Message:
FileNotFoundError: No such file or directory '/tmp/tmpa_7dek7j/t1-linear/8bd0d87777eff53210d9b9af51596e2965918279/n4biascorrection/sub-001_ses-01_T1w_bias.nii.gz' for output 'bias_image' of a N4BiasFieldCorrection interface
Environment:

Linux OS
Python 3.8
Clinica installed via Miniconda
I have verified that the input file /home/zzc/Output/sub-001/ses-01/anat/sub-001_ses-01_T1w.nii.gz exists and is readable. However, the output 'bias_image' file does not seem to get generated, leading to the error.

Could you provide any insights or solutions to this issue? Your assistance would be greatly appreciated.

Best regards

**Comment by NicolasGensollen** [comment]
Hi @mtmmu88 

Thanks for reporting. 
Could you give us some more details about the input data (a `tree` of your input folder or something equivalent would be great).

You can also try to see if Clinica is able to find your files by using the "low level" python function. You basically need to open a Python terminal and run the following:

```python
>>> from clinica.utils.inputs import clinica_file_reader, check_bids_folder
>>> from clinica.utils.participant import get_subject_session_list
>>> from clinica.utils.input_files import T1W_NII
>>> bids_dir = "/home/zzc/Output"
>>> check_bids_folder(bids_dir)
>>> sessions, subjects = get_subject_session_list(bids_dir)
>>> sessions # check the sessions
>>> subjects # check the subjects
>>> files, errors = clinica_file_reader(subjects, sessions, bids_dir, T1W_NII)
>>> files # check the files found
>>> errors # check the list of errors
```

If everything went well, you should have a list of paths to T1W files for the variable `files` and errors should be `'Clinica encountered 0 problem(s) while getting T1w MRI:\n'`

If the file reader is able to find your input T1w images, we'll take a closer look at what is happening later in the pipeline.

Let us know how it went.
Best,
Nicolas

**Comment by mtmmu88** [comment]
Hi Nicolas,
Thank you for your swift response. I followed the diagnostic steps you
provided and everything seems fine on that end. Below are the details of
the code I executed and its outputs:
from clinica.utils.inputs import clinica_file_reader, check_bids_folder
from clinica.utils.participant import get_subject_session_list
from clinica.utils.input_files import T1W_NII

bids_dir = "/home/zzc/Output"
check_bids_folder(bids_dir)
sessions, subjects = get_subject_session_list(bids_dir)

# Checking sessions and subjects
print(sessions)  # ['ses-01', 'ses-01', 'ses-01', 'ses-01', 'ses-01']
print(subjects)  # ['sub-001', 'sub-002', 'sub-003', 'sub-004', 'sub-005']

files, errors = clinica_file_reader(subjects, sessions, bids_dir, T1W_NII)

# Checking the files found and errors
print(files)  #
['/home/zzc/Output/sub-001/ses-01/anat/sub-001_ses-01_T1w.nii.gz',
'/home/zzc/Output/sub-002/ses-01/anat/sub-002_ses-01_T1w.nii.gz',
'/home/zzc/Output/sub-003/ses-01/anat/sub-003_ses-01_T1w.nii.gz',...]
print(errors)  # 'Clinica encountered 0 problem(s) while getting T1w MRI:\n'

As you can see, Clinica is able to find the input T1w images in my BIDS
directory, and there were no problems encountered during this stage. I'd
appreciate it if you could take a closer look at what might be happening
later in the pipeline.

Best regards,

mtmmu88

Gensollen ***@***.***> 于2023年10月17日周二 17:28写道：

**Comment by NicolasGensollen** [comment]
@mtmmu88 Thanks for trying this out. Indeed, it seems like the file reader is successful at finding your input files which is a good sign.

At this point, I am not quite sure why you get this error. It could be an issue with ANTS or its installation on your system.
One thing that you can try is to run the ANTS command directly in your terminal to see if it is successful or not.

```
$ N4BiasFieldCorrection --bspline-fitting [ 600 ] -d 3 --input-image /home/zzc/Output/sub-001/ses-01/anat/sub-001_ses-01_T1w.nii.gz --output [ sub-001_ses-01_T1w_corrected.nii.gz, sub-001_ses-01_T1w_bias.nii.gz ]
```

After running this command, you should have two new non-empty files in the folder from which you ran the command above:
- `sub-001_ses-01_T1w_corrected.nii.gz`
- `sub-001_ses-01_T1w_bias.nii.gz`

**Comment by mtmmu88** [comment]
Hi Nicolas,

Thank you for your guidance. I've attempted to run the
N4BiasFieldCorrection command directly in my terminal on two different
systems:

   1.

   On an 8-core, 8GB RAM Linux system, the command with --bspline-fitting [
   600 ] was killed. However, when I reduced the bspline-fitting parameter
   to 300, the command executed successfully and generated the two expected
   files.
   2.

   On a 12-core, 40GB RAM system, the command with --bspline-fitting [ 600 ]
   also executed successfully and produced the expected files.

Despite these tests, I still encounter the original issue when running
Clinica's t1-linear pipeline on both systems.

Given these new findings, could you offer additional insights or potential
solutions?

Best regards,

mtmmu88

Gensollen ***@***.***> 于2023年10月17日周二 23:32写道：

**Comment by NicolasGensollen** [comment]
Hi @mtmmu88 

Thanks a lot for trying these out. I'm not sure to understand what's happening on your machines. I initially thought that you were having a RAM issue on your 8GB system which was causing the `N4BiasFieldCorrection` step to fail, but you shouldn't get the exact same error on the 40 GB system.

Could you take a look at what files were written in the `tmp` folder after trying to run the pipeline (for example run a `$ tree /tmp/tmpXXXXX` where XXXX is the hash you can find in the error message.

Could you also report the full error trace ? 

Also, which version of ANTS are you using ?
 
Besides, I tried mimicking your settings on a Linux machine:

```
$ tree
.
├── bids
│   ├── sub-001
│   │   └── ses-01
│   │       └── anat
│   │           └── sub-001_ses-01_T1w.nii.gz
│   ├── sub-002
│   │   └── ses-01
│   │       └── anat
│   │           └── sub-002_ses-01_T1w.nii.gz
│   └── sub-003
│       └── ses-01
│           └── anat
│               └── sub-003_ses-01_T1w.nii.gz
└── caps
```

And was able to run the pipeline:

```
$ clinica run t1-linear ./bids ./caps
2023-10-18 10:53:18,957:WARNING:Could not determine if /builds/test_t1_linear/caps is a CAPS or BIDS directory. Clinica will assume this is a CAPS directory.
2023-10-18 10:53:18,985:INFO:The pipeline will be run on the following 3 image(s):
2023-10-18 10:53:18,985:INFO:	sub-001 | ses-01,
2023-10-18 10:53:18,985:INFO:	sub-002 | ses-01,
2023-10-18 10:53:18,986:INFO:	sub-003 | ses-01,
2023-10-18 10:53:18,986:INFO:The pipeline will last approximately 6 minutes per image.
2023-10-18 10:53:19,046:WARNING:You did not specify the number of threads to run in parallel (--n_procs argument).
2023-10-18 10:53:19,046:WARNING:Computation time can be shorten as you have 4 CPUs on this computer. We recommend using 3 threads.
How many threads do you want to use? [3]:
Number of threads set to 3. You may set the --n_procs argument to disable this message for future calls.
2023-10-18 10:57:31,545:INFO:sub-001_ses-01 has completed
2023-10-18 10:57:33,544:INFO:sub-002_ses-01 has completed
2023-10-18 10:57:39,550:INFO:sub-003_ses-01 has completed
2023-10-18 10:57:43,566:INFO:The t1-linear pipeline has completed. Working directory was automatically deleted.
```

After this you should have:

```
$ tree
.
├── bids
│   ├── sub-001
│   │   └── ses-01
│   │       └── anat
│   │           └── sub-001_ses-01_T1w.nii.gz
│   ├── sub-002
│   │   └── ses-01
│   │       └── anat
│   │           └── sub-002_ses-01_T1w.nii.gz
│   └── sub-003
│       └── ses-01
│           └── anat
│               └── sub-003_ses-01_T1w.nii.gz
├── caps
│   └── subjects
│       ├── sub-001
│       │   └── ses-01
│       │       └── t1_linear
│       │           ├── sub-001_ses-0_space-MNI152NLin2009cSym_desc-Crop_res-1x1x1_T1w.nii.gz
│       │           ├── sub-001_ses-0_space-MNI152NLin2009cSym_res-1x1x1_T1w.nii.gz
│       │           └── sub-001_ses-0_space-MNI152NLin2009cSym_res-1x1x1_affine.mat
│       ├── sub-002
│       │   └── ses-01
│       │       └── t1_linear
│       │           ├── sub-002_ses-0_space-MNI152NLin2009cSym_desc-Crop_res-1x1x1_T1w.nii.gz
│       │           ├── sub-002_ses-0_space-MNI152NLin2009cSym_res-1x1x1_T1w.nii.gz
│       │           └── sub-002_ses-0_space-MNI152NLin2009cSym_res-1x1x1_affine.mat
│       └── sub-003
│           └── ses-01
│               └── t1_linear
│                   ├── sub-003_ses-0_space-MNI152NLin2009cSym_desc-Crop_res-1x1x1_T1w.nii.gz
│                   ├── sub-003_ses-0_space-MNI152NLin2009cSym_res-1x1x1_T1w.nii.gz
│                   └── sub-003_ses-0_space-MNI152NLin2009cSym_res-1x1x1_affine.mat
└── pypeline.log
```

There seems to be a small issue with the output session labels in the filenames (usually we use the format `ses-MXXX` for session labels) but that shouldn't cause the issue you're having.

**Comment by ghisvail** [comment]
Just out of curiosity, what is the size of the data you are trying to process, i.e. dimensionality of each volume and the total number of volumes (normally subjects x sessions)?

**Comment by NicolasGensollen** [comment]
Adding on @ghisvail's comments, one thing you can also try is to set the number of parallel jobs to 1 to see if you face the same issue : 

```
$ clinica run t1-linear BIDS_FOLDER CAPS_FOLDER --n_procs=1
```

If you don't experience the same issue, you can then try to increase this number to speed up your processing.

**Comment by mtmmu88** [comment]
Thank you for your continued support. I have been diligently following the
steps you've provided. Unfortunately, the issue persists.
1.The temporary directory structure is as shown:
tree /tmp/tmpblxkm_u7
/tmp/tmpblxkm_u7
└── t1-linear
    ├── 19cf7e01662661485f20dc61f8f4e860484493e6
    │   ├── ImageID
    │   │   ├── _0x8664a92d3b7f2f8d3698fb363ab69434.json
    │   │   ├── _inputs.pklz
    │   │   ├── _node.pklz
    │   │   ├── _report
    │   │   │   └── report.rst
    │   │   └── result_ImageID.pklz
    │   └── n4biascorrection
    │       ├── _inputs.pklz
    │       ├── _node.pklz
    │       ├── _report
    │       │   └── report.rst
    │       ├── command.txt
    │       └── result_n4biascorrection.pklz
    ├── d3.js
    ├── graph.json
    ├── graph1.json
    └── index.html

6 directories, 14 files
2.I successfully ran the ANTs N4BiasFieldCorrection command manually on two
different systems. The 40GB RAM system processed the command successfully,
but the 8GB RAM system did not.
3.I am conducting generalizability tests with clinical research data from a
hospital.My data has a dimensionality of (156, 512, 512), and I have a
total of 5 subjects.
4.I'm using ANTs version 2.4.4 compiled on May 1, 2023.

I kindly ask for further guidance in resolving this issue. Your tool is
vital for my research, and it is imperative to get it running smoothly.

Please find my pypeline.log attached for your reference.

Let me know if more information is required from my end.

Best regards,

mtmmu88

Attachment: pypeline.log Feel free to use this revised template for your
email.

Ghislain Vaillant ***@***.***> 于2023年10月18日周三 17:40写道：

**Comment by mtmmu88** [comment]
I've tried running the pipeline with --n_procs=1 to limit the number of
parallel jobs, but the issue persists on both my 8GB and 40GB systems.

Gensollen ***@***.***> 于2023年10月18日周三 17:54写道：

**Comment by ghisvail** [comment]
> Attachment: pypeline.log

Could you provide your `pypeline.log` corresponding to your latest try? Looks like GitHub discards it out when attaching the file by email.

**Comment by mtmmu88** [comment]
I've uploaded the pypeline.log to GitHub for your review.

Ghislain Vaillant ***@***.***> 于2023年10月18日周三 19:45写道：

**Comment by mtmmu88** [comment]
Hi Nicolas,

I have been working on resolving issues with the clinica run t1-linear
pipeline. After implementing the solution for issue #1000 from the
project's issue tracker, the FileNotFoundError for N4BiasFieldCorrection
has been resolved. However, a new issue has emerged concerning a mismatch
between the expected and actual bytes while reading the NIfTI file. Despite
verifying the integrity of the file using a NIfTI viewer, the problem
persists.

This issue is causing significant delays in our project timeline, and I
would greatly appreciate your prompt assistance in resolving this matter.

Thank you for your attention to this issue.

Best regards,

mtmmu88


mtmmu ***@***.***> 于2023年10月18日周三 20:17写道：

**Comment by NicolasGensollen** [comment]
Hi @mtmmu88 

I don't think #1000 is fixing the issue you were encountering. It is only fixing a file name formatting issue.

I might be wrong, but I think your issue comes from a lack of RAM which causes spurious errors at various stages of the pipeline. Usually, the consequence is that the next node is missing part of its inputs which triggers a `FileNotFound` error.

Could you share the new error that you got ?

**Comment by mtmmu88** [comment]
Hi，

Thank you for your continued support. I have followed the steps outlined in
issue #1000, but a new issue has arisen, indicating file corruption. I have
successfully processed data from the ADNI public database on both an 8GB
and a 40GB RAM machine without any issues. However, when using my own data,
which has been organized according to BIDS standards and validated, the
same problem recurs on both machines.

I am attaching the new error log for your review. Given the urgency of
resolving this issue for my ongoing research, I would greatly appreciate
your prompt attention.

Gensollen ***@***.***> 于2023年10月20日周五 16:56写道：

**Comment by NicolasGensollen** [comment]
Hi @mtmmu88 

Did you run this pipeline without setting the number of cores ? 
Having a quick look at the logs you provided, the pipeline seems to be running on 8 cores and I think this is causing RAM issues. Given your memory / number of cores ratio, I would limit the number of jobs to 1.

**Comment by mtmmu88** [comment]
Thank you for your latest response and suggestion to limit the number of
cores to one for mitigating RAM issues. I have already tried this approach,
but the issue still persists.

Interestingly, I don't encounter these issues when using data downloaded
from the ADNI database. The problem only arises when I use my own data,
which has been organized according to BIDS standards and verified through
the BIDS validator.

Could you please advise on further steps to resolve this puzzling issue?


Gensollen ***@***.***> 于2023年10月20日周五 19:24写道：

**Comment by mtmmu88** [comment]
Dear Clinica Team,

I wanted to update you on the issue I was facing. After uninstalling
Clinica and deleting the conda environment, then reinstalling everything
and reorganizing my data according to BIDS standards, the issue has been
resolved.

However, I still encounter the N4 node issue on machines with lower memory,
which leads me to believe it might be a RAM-related issue after all.

Thank you for your assistance throughout this process.

Best regards,

mtmmu88

mtmmu ***@***.***> 于2023年10月20日周五 19:57写道：
